### ISSUE TITLE
[SCM-19] resolve file status ambiguities

### ISSUE THREAD

**Issue by olamy** [olamy/maven-scm#9]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** opened **[SCM-19](https://issues.apache.org/jira/browse/SCM-19?redirect=false)** and commented

currently ScmFileStatus just reports whatever the provider gave back, but means different things in different SCM's.

eg. "updated" in CVS is UPDATED or PATCHED, in SVN it is UPDATED or MERGED.
"added" in CVS is UPDATED, in SVN it is ADDED.

The API should hide this from the user, though perhaps be able to give the extended info if necessary.

I suggest isXXX flags instead:
isUpdated()
isAdded() - just came from an update
isMerged()
isInConflict()
isNew() - new locally
etc.

---

**Attachments:**
- [MNG-19_1-maven-scm-api.patch](https://issues.apache.org/jira/secure/attachment/12718686/MNG-19_1-maven-scm-api.patch) (_4.37 kB_)
- [MNG-19-maven-scm-api.patch](https://issues.apache.org/jira/secure/attachment/12718489/MNG-19-maven-scm-api.patch) (_3.50 kB_)

**Comment by olamy** [comment]
**[Torbjorn Eikli Smorgrav](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=smorgrav)** commented

I think the ScmFileStatus should be independent of the providers.
It should define its own semantics.
The CVS provider should return added when a file is added.
And I think patched is redundant.

I have a related problem working with Bazaar and I need a method that indicates if
the status indicates some kind of change eg. hasDiff().

I have made a patch where I tried to comment each status and added that hasDiff method to
show you what I mean.

Take a look

**Comment by olamy** [comment]
**[Torbjorn Eikli Smorgrav](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=smorgrav)** commented

I realized that I was not very consistent between my patch and my comment above.
To clearify:

Some of the status are related to the status of the file in the working tree versus the repository eg. added and removed.

Others are just a tag that the file is involved in a transaction with the repository eg. checkedin and checkout.

Some status eg. update, added and removed are used in both cases and I think it should be separated.
Updated should In my opinion be a transaction tag.

T
